docs/import-cli.rst

```diff
@@ -260,6 +264,59 @@ Note: Connecting over SSH is necessary only if you intend to import in-place. If
+
+Further, the images you are importing might need a range or different rendering settings, not just one set of settings for all of them. Also for this case, the step-by-step approach, first importing the images, only then deciding on the rendering strategy and preparing the `renderingdef.yml` files, is preferable.
+
+Nevertheless, there are cases which do not need visual checks and use a single rendering for all images, for which a streamlined sequence of commands is offered below which will perform all three steps (import, rendering and metadata import) in one single session on the CLI.
+
+Resources
+---------
+
+Additionally to the :ref:`Resources mentioned in the import-cli section<Resources>` and in the :ref:`Setup<Setup>` you will also need the rendering and metadata plugins as mentioned in :doc:`render` and :doc:`metadata`, and possibly the following files:
+
+   - https://github.com/ome/training-scripts/blob/master/maintenance/preparation/renderingdef.yml
+   - :download:`simple-annotation.csv <../scripts/simple-annotation.csv>`
+   - :download:`simple-annotation-bulkmap-config.yml <../scripts/simple-annotation-bulkmap-config.yml>`
+
+Step-by-step
+------------
+
+#. If you did not do so already, open a terminal window on your local machine and activate the conda environment where your ``omero-py`` is installed (see the :ref:`Setup<Setup>`)::
+
+   $ conda activate myenv
+
+#. Set the ``OMERODIR`` variable to point to the downloaded and unzipped OMERO server dir (see the :ref:`Setup<Setup>`)::
+
+   $ export OMERODIR=/path/to/OMERO.server-x.x.x-ice36-bxx
+
+#. Prepare an `renderingdef.yml` file, either creating a new one or downloading https://github.com/ome/training-scripts/blob/master/maintenance/preparation/renderingdef.yml.
+
+#. Prepare an `annotation.csv` file, by creating a new file or downloading the provided example file. In the example below, we use the file :download:`simple-annotation.csv <../scripts/simple-annotation.csv>`. The Dataset names in this `CSV` file must match the Dataset names in OMERO as created in the `DID` variable definition line in the command below. The Image names in the `CSV` file must match the file names in your imported folder.
+
+#. Prepare a `bulkmap-config.yml` file. In the example below, we use the file :download:`simple-annotation-bulkmap-config.yml <../scripts/simple-annotation-bulkmap-config.yml>`.
+
+#. Log in to the OMERO.server you wish to import to. This can be a remote server if you do not wish to import `in-place`.
+
+#. Import and render in a single command sequence below::
+
+   $  PID=$(omero obj new Project name='Project_import_concatenate')
+   $  DID=$(omero obj new Dataset name='siRNAi-HeLa')
+   $  omero obj new ProjectDatasetLink parent=$PID child=$DID
+   $  omero import -d $DID /path/to/data/folder/or/image/siRNAi-HeLa --file import.out
+   $  for i in $(cat import.out); do omero render set $i renderingdef.yml; done
```

Comment:
`import.out` is the full log output
```
15:03:48.289 [main] INFO  ome.formats.importer.ImportConfig - OMERO.blitz Version: 5.5.7
15:03:48.314 [main] INFO  ome.formats.importer.ImportConfig - Bioformats version: 6.5.1 revision: 6f50e4d52c9d96112635fd8b2dde737f31041cf0 date: 7 July 2020
15:03:48.381 [main] INFO  o.f.importer.cli.CommandLineImporter - Log levels -- Bio-Formats: null OMERO.importer: null
15:03:48.382 [main] DEBUG ome.formats.importer.ImportConfig - Loaded savedDirectory from ome.formats.importer.util.IniFileLoader@4de8b406
15:03:48.382 [main] DEBUG ome.formats.importer.ImportConfig - Loaded savedDirectory from default
15:03:48.402 [main] DEBUG loci.formats.ClassList - Could not find loci.formats.in.URLReader
...
Image:134289
...
```

Reply:
You can ignore this, it'll be fixed by https://github.com/ome/openmicroscopy/pull/6255

Reply:
> import.out is the full log output

Thanks. Interestingly, it was not for me during my testing. I was getting the output only such as ``Image:ID1`` etc. And thus the workflow worked for me as described.
What is the difference between our testing envs I wonder @manics ?

Reply:
I'm using the new auto-download jars feature in OMERO.py 5.8.0

Reply:
```
(env562) pwalczysko@ls31618~$ echo $OMERODIR 
/Users/pwalczysko/Downloads/OMERO.server-5.6.1-ice36-b225
(env562) pwalczysko@ls31618~$ pip freeze | grep omero-py
omero-py==5.8.0
(env562) pwalczysko@ls31618~$ omero import  ~/testing --file import-test.out
Using session for trainer-1@outreach.openmicroscopy.org:4064. Idle timeout: 10 min. Current group: Lab1
2020-09-24 17:47:31,559 254        [      main] INFO          ome.formats.importer.ImportConfig - OMERO.blitz Version: 5.5.6
2020-09-24 17:47:31,574 269        [      main] INFO          ome.formats.importer.ImportConfig - Bioformats version: 6.3.1 revision: e355a853c5768721711cb9b29c3a27561539c4e4 date: 13 December 2019
2020-09-24 17:47:31,614 309        [      main] INFO   formats.importer.cli.CommandLineImporter - Log levels -- Bio-Formats: ERROR OMERO.importer: INFO
2020-09-24 17:47:31,965 660        [      main] INFO      ome.formats.importer.ImportCandidates - Depth: 4 Metadata Level: MINIMUM
2020-09-24 17:47:32,024 719        [      main] INFO      ome.formats.importer.ImportCandidates - 2 file(s) parsed into 2 group(s) with 2 call(s) to setId in 55ms. (58ms total) [0 unknowns]
2020-09-24 17:47:32,063 758        [      main] INFO       ome.formats.OMEROMetadataStoreClient - Attempting initial SSL connection to outreach.openmicroscopy.org:4064
2020-09-24 17:47:32,973 1668       [      main] INFO       ome.formats.OMEROMetadataStoreClient - Insecure connection requested, falling back
2020-09-24 17:47:34,488 3183       [      main] INFO       ome.formats.OMEROMetadataStoreClient - Pinging session every 300s.
2020-09-24 17:47:34,622 3317       [      main] INFO       ome.formats.OMEROMetadataStoreClient - Server: 5.6.2
2020-09-24 17:47:34,622 3317       [      main] INFO       ome.formats.OMEROMetadataStoreClient - Client: 5.5.6
2020-09-24 17:47:34,622 3317       [      main] INFO       ome.formats.OMEROMetadataStoreClient - Java Version: 1.8.0_222
2020-09-24 17:47:34,622 3317       [      main] INFO       ome.formats.OMEROMetadataStoreClient - OS Name: Mac OS X
2020-09-24 17:47:34,622 3317       [      main] INFO       ome.formats.OMEROMetadataStoreClient - OS Arch: x86_64
2020-09-24 17:47:34,622 3317       [      main] INFO       ome.formats.OMEROMetadataStoreClient - OS Version: 10.13.6
2020-09-24 17:47:35,334 4029       [      main] WARN                    ome.system.UpgradeCheck - UPGRADE AVAILABLE:Please upgrade to 5.6.1. See http://downloads.openmicroscopy.org/latest/omero for the latest version.
2020-09-24 17:47:35,345 4040       [2-thread-1] INFO   ormats.importer.cli.LoggingImportMonitor - FILESET_UPLOAD_PREPARATION
2020-09-24 17:47:35,687 4382       [2-thread-1] INFO   ormats.importer.cli.LoggingImportMonitor - FILESET_UPLOAD_START
2020-09-24 17:47:35,782 4477       [3-thread-1] INFO   ts.importer.transfers.UploadFileTransfer - Transferring /Users/pwalczysko/testing/HDF_logo.svg.png...
2020-09-24 17:47:35,840 4535       [3-thread-1] INFO   ormats.importer.cli.LoggingImportMonitor - FILE_UPLOAD_STARTED: /Users/pwalczysko/testing/HDF_logo.svg.png
2020-09-24 17:47:36,132 4827       [3-thread-1] INFO   ormats.importer.cli.LoggingImportMonitor - FILE_UPLOAD_COMPLETE: /Users/pwalczysko/testing/HDF_logo.svg.png
2020-09-24 17:47:36,309 5004       [2-thread-1] INFO   ormats.importer.cli.LoggingImportMonitor - FILESET_UPLOAD_END
2020-09-24 17:47:36,411 5106       [l.Client-0] INFO   ormats.importer.cli.LoggingImportMonitor - METADATA_IMPORTED Step: 1 of 5  Logfile: null
2020-09-24 17:47:36,634 5329       [l.Client-1] INFO   ormats.importer.cli.LoggingImportMonitor - PIXELDATA_PROCESSED Step: 2 of 5  Logfile: null
2020-09-24 17:47:36,700 5395       [2-thread-1] INFO   ormats.importer.cli.LoggingImportMonitor - IMPORT_STARTED Logfile: 289330
2020-09-24 17:47:36,761 5456       [l.Client-1] INFO   ormats.importer.cli.LoggingImportMonitor - THUMBNAILS_GENERATED Step: 3 of 5  Logfile: 289330
2020-09-24 17:47:36,777 5472       [l.Client-0] INFO   ormats.importer.cli.LoggingImportMonitor - METADATA_PROCESSED Step: 4 of 5  Logfile: 289330
2020-09-24 17:47:36,789 5484       [l.Client-1] INFO   ormats.importer.cli.LoggingImportMonitor - OBJECTS_RETURNED Step: 5 of 5  Logfile: 289330
2020-09-24 17:47:36,930 5625       [l.Client-0] INFO   ormats.importer.cli.LoggingImportMonitor - IMPORT_DONE Imported file: /Users/pwalczysko/testing/HDF_logo.svg.png
Other imported objects:
Fileset:42546
2020-09-24 17:47:36,991 5686       [2-thread-1] INFO   ormats.importer.cli.LoggingImportMonitor - FILESET_UPLOAD_PREPARATION
2020-09-24 17:47:37,253 5948       [2-thread-1] INFO   ormats.importer.cli.LoggingImportMonitor - FILESET_UPLOAD_START
2020-09-24 17:47:37,342 6037       [3-thread-1] INFO   ts.importer.transfers.UploadFileTransfer - Transferring /Users/pwalczysko/testing/HDR5-2.png...
2020-09-24 17:47:37,402 6097       [3-thread-1] INFO   ormats.importer.cli.LoggingImportMonitor - FILE_UPLOAD_STARTED: /Users/pwalczysko/testing/HDR5-2.png
2020-09-24 17:47:37,677 6372       [3-thread-1] INFO   ormats.importer.cli.LoggingImportMonitor - FILE_UPLOAD_COMPLETE: /Users/pwalczysko/testing/HDR5-2.png
2020-09-24 17:47:37,847 6542       [2-thread-1] INFO   ormats.importer.cli.LoggingImportMonitor - FILESET_UPLOAD_END
2020-09-24 17:47:38,087 6782       [2-thread-1] INFO   ormats.importer.cli.LoggingImportMonitor - IMPORT_STARTED Logfile: 289333
2020-09-24 17:47:38,130 6825       [l.Client-1] INFO   ormats.importer.cli.LoggingImportMonitor - PIXELDATA_PROCESSED Step: 2 of 5  Logfile: 289333
2020-09-24 17:47:38,253 6948       [l.Client-0] INFO   ormats.importer.cli.LoggingImportMonitor - THUMBNAILS_GENERATED Step: 3 of 5  Logfile: 289333
2020-09-24 17:47:38,269 6964       [l.Client-1] INFO   ormats.importer.cli.LoggingImportMonitor - METADATA_PROCESSED Step: 4 of 5  Logfile: 289333
2020-09-24 17:47:38,280 6975       [l.Client-0] INFO   ormats.importer.cli.LoggingImportMonitor - OBJECTS_RETURNED Step: 5 of 5  Logfile: 289333
2020-09-24 17:47:38,383 7078       [l.Client-1] INFO   ormats.importer.cli.LoggingImportMonitor - IMPORT_DONE Imported file: /Users/pwalczysko/testing/HDR5-2.png
Other imported objects:
Fileset:42547

==> Summary
2 files uploaded, 2 filesets created, 2 images imported, 0 errors in 0:00:03.092
(env562) pwalczysko@ls31618~$ code import-test.out

```

which shows the output file with 

```
Image:58694
Image:58695

```

Maybe I do not understand how to use the auto-download feature of OMERO.py 580 ?

Reply:
Aha, okay, sorry, understood. I imported now with no OMERODIR set and some download of jars happened for me.
Now I can see that also for me, there is a lot of unwanted output in the ``import.out`` file, as well as two exceptions.

```
(env562) pwalczysko@ls31618~$ OMERODIR=
(env562) pwalczysko@ls31618~$ echo $OMERODIR 

(env562) pwalczysko@ls31618~$ 
(env562) pwalczysko@ls31618~$ 
(env562) pwalczysko@ls31618~$ 
(env562) pwalczysko@ls31618~$ omero import  ~/testing --file import-test.out
Downloading https://downloads.openmicroscopy.org/omero/latest/OMERO.java.zip
Using OMERO.java-5.6.2-ice36-b226
Using session for trainer-1@outreach.openmicroscopy.org:4064. Idle timeout: 10 min. Current group: Lab1
Other imported objects:
Fileset:42548
Other imported objects:
Fileset:42549

==> Summary
2 files uploaded, 2 filesets created, 2 images imported, 0 errors in 0:00:03.089
(env562) pwalczysko@ls31618~$ 
```

And the ``iimport.out`` contains

<details>

```
17:55:52.265 [main] INFO  ome.formats.importer.ImportConfig - OMERO.blitz Version: 5.5.7
17:55:52.284 [main] INFO  ome.formats.importer.ImportConfig - Bioformats version: 6.5.1 revision: 6f50e4d52c9d96112635fd8b2dde737f31041cf0 date: 7 July 2020
17:55:52.326 [main] INFO  o.f.importer.cli.CommandLineImporter - Log levels -- Bio-Formats: null OMERO.importer: null
17:55:52.328 [main] DEBUG ome.formats.importer.ImportConfig - Loaded savedDirectory from ome.formats.importer.util.IniFileLoader@4de8b406
17:55:52.328 [main] DEBUG ome.formats.importer.ImportConfig - Loaded savedDirectory from default
17:55:52.344 [main] DEBUG loci.formats.ClassList - Could not find loci.formats.in.URLReader
java.lang.ClassNotFoundException: loci.formats.in.URLReader
	at java.net.URLClassLoader.findClass(URLClassLoader.java:382) ~[na:1.8.0_222]
	at java.lang.ClassLoader.loadClass(ClassLoader.java:424) ~[na:1.8.0_222]
	at sun.misc.Launcher$AppClassLoader.loadClass(Launcher.java:349) ~[na:1.8.0_222]
	at java.lang.ClassLoader.loadClass(ClassLoader.java:357) ~[na:1.8.0_222]
	at java.lang.Class.forName0(Native Method) ~[na:1.8.0_222]
	at java.lang.Class.forName(Class.java:264) ~[na:1.8.0_222]
	at loci.formats.ClassList.parseLine(ClassList.java:196) [formats-api.jar:6.5.1]
	at loci.formats.ClassList.parseFile(ClassList.java:258) [formats-api.jar:6.5.1]
	at loci.formats.ClassList.<init>(ClassList.java:138) [formats-api.jar:6.5.1]
	at loci.formats.ClassList.<init>(ClassList.java:122) [formats-api.jar:6.5.1]
	at loci.formats.ImageReader.getDefaultReaderClasses(ImageReader.java:80) [formats-api.jar:6.5.1]
	at ome.formats.importer.OMEROWrapper.createReader(OMEROWrapper.java:141) [omero-blitz.jar:5.5.7]
	at ome.formats.importer.OMEROWrapper.<init>(OMEROWrapper.java:103) [omero-blitz.jar:5.5.7]
	at ome.formats.importer.OMEROWrapper.<init>(OMEROWrapper.java:95) [omero-blitz.jar:5.5.7]
	at ome.formats.importer.OMEROWrapper.<init>(OMEROWrapper.java:90) [omero-blitz.jar:5.5.7]
	at ome.formats.importer.cli.CommandLineImporter.<init>(CommandLineImporter.java:142) [omero-blitz.jar:5.5.7]
	at ome.formats.importer.cli.CommandLineImporter.main(CommandLineImporter.java:997) [omero-blitz.jar:5.5.7]
17:55:52.352 [main] DEBUG loci.formats.ClassList - Could not find loci.formats.in.SlideBook6Reader
java.lang.ClassNotFoundException: loci.formats.in.SlideBook6Reader
	at java.net.URLClassLoader.findClass(URLClassLoader.java:382) ~[na:1.8.0_222]
	at java.lang.ClassLoader.loadClass(ClassLoader.java:424) ~[na:1.8.0_222]
	at sun.misc.Launcher$AppClassLoader.loadClass(Launcher.java:349) ~[na:1.8.0_222]
	at java.lang.ClassLoader.loadClass(ClassLoader.java:357) ~[na:1.8.0_222]
	at java.lang.Class.forName0(Native Method) ~[na:1.8.0_222]
	at java.lang.Class.forName(Class.java:264) ~[na:1.8.0_222]
	at loci.formats.ClassList.parseLine(ClassList.java:196) [formats-api.jar:6.5.1]
	at loci.formats.ClassList.parseFile(ClassList.java:258) [formats-api.jar:6.5.1]
	at loci.formats.ClassList.<init>(ClassList.java:138) [formats-api.jar:6.5.1]
	at loci.formats.ClassList.<init>(ClassList.java:122) [formats-api.jar:6.5.1]
	at loci.formats.ImageReader.getDefaultReaderClasses(ImageReader.java:80) [formats-api.jar:6.5.1]
	at ome.formats.importer.OMEROWrapper.createReader(OMEROWrapper.java:141) [omero-blitz.jar:5.5.7]
	at ome.formats.importer.OMEROWrapper.<init>(OMEROWrapper.java:103) [omero-blitz.jar:5.5.7]
	at ome.formats.importer.OMEROWrapper.<init>(OMEROWrapper.java:95) [omero-blitz.jar:5.5.7]
	at ome.formats.importer.OMEROWrapper.<init>(OMEROWrapper.java:90) [omero-blitz.jar:5.5.7]
	at ome.formats.importer.cli.CommandLineImporter.<init>(CommandLineImporter.java:142) [omero-blitz.jar:5.5.7]
	at ome.formats.importer.cli.CommandLineImporter.main(CommandLineImporter.java:997) [omero-blitz.jar:5.5.7]
17:55:52.353 [main] DEBUG loci.formats.ClassList - Could not find loci.formats.in.ScreenReader
java.lang.ClassNotFoundException: loci.formats.in.ScreenReader
	at java.net.URLClassLoader.findClass(URLClassLoader.java:382) ~[na:1.8.0_222]
	at java.lang.ClassLoader.loadClass(ClassLoader.java:424) ~[na:1.8.0_222]
	at sun.misc.Launcher$AppClassLoader.loadClass(Launcher.java:349) ~[na:1.8.0_222]
	at java.lang.ClassLoader.loadClass(ClassLoader.java:357) ~[na:1.8.0_222]
	at java.lang.Class.forName0(Native Method) ~[na:1.8.0_222]
	at java.lang.Class.forName(Class.java:264) ~[na:1.8.0_222]
	at loci.formats.ClassList.parseLine(ClassList.java:196) [formats-api.jar:6.5.1]
	at loci.formats.ClassList.parseFile(ClassList.java:258) [formats-api.jar:6.5.1]
	at loci.formats.ClassList.<init>(ClassList.java:138) [formats-api.jar:6.5.1]
	at loci.formats.ClassList.<init>(ClassList.java:122) [formats-api.jar:6.5.1]
	at loci.formats.ImageReader.getDefaultReaderClasses(ImageReader.java:80) [formats-api.jar:6.5.1]
	at ome.formats.importer.OMEROWrapper.createReader(OMEROWrapper.java:141) [omero-blitz.jar:5.5.7]
	at ome.formats.importer.OMEROWrapper.<init>(OMEROWrapper.java:103) [omero-blitz.jar:5.5.7]
	at ome.formats.importer.OMEROWrapper.<init>(OMEROWrapper.java:95) [omero-blitz.jar:5.5.7]
	at ome.formats.importer.OMEROWrapper.<init>(OMEROWrapper.java:90) [omero-blitz.jar:5.5.7]
	at ome.formats.importer.cli.CommandLineImporter.<init>(CommandLineImporter.java:142) [omero-blitz.jar:5.5.7]
	at ome.formats.importer.cli.CommandLineImporter.main(CommandLineImporter.java:997) [omero-blitz.jar:5.5.7]
17:55:52.670 [main] DEBUG o.s.nativelib.NativeLibraryUtil - processor is INTEL_64 os.arch is x86_64
17:55:52.670 [main] DEBUG o.s.nativelib.NativeLibraryUtil - architecture is OSX_64 os.name is mac os x
17:55:52.672 [main] DEBUG o.s.nativelib.NativeLibraryUtil - architecture is OSX_64 os.name is mac os x
17:55:52.672 [main] DEBUG o.s.nativelib.NativeLibraryUtil - platform specific path is META-INF/lib/osx_64/
17:55:52.672 [main] DEBUG o.scijava.nativelib.BaseJniExtractor - mappedLib is libturbojpeg.dylib
17:55:52.672 [main] DEBUG o.scijava.nativelib.BaseJniExtractor - URL is jar:file:/Users/pwalczysko/Library/Caches/OMERO.py/5/jars/OMERO.java-5.6.2-ice36-b226/libs/turbojpeg.jar!/META-INF/lib/osx_64/libturbojpeg.dylib
17:55:52.672 [main] DEBUG o.scijava.nativelib.BaseJniExtractor - URL path is file:/Users/pwalczysko/Library/Caches/OMERO.py/5/jars/OMERO.java-5.6.2-ice36-b226/libs/turbojpeg.jar!/META-INF/lib/osx_64/libturbojpeg.dylib
17:55:52.674 [main] DEBUG o.scijava.nativelib.BaseJniExtractor - Extracting 'jar:file:/Users/pwalczysko/Library/Caches/OMERO.py/5/jars/OMERO.java-5.6.2-ice36-b226/libs/turbojpeg.jar!/META-INF/lib/osx_64/libturbojpeg.dylib' to '/var/folders/j3/9sf00sjs4kng7djn76m0_grh4d75r0/T/libturbojpeg7980512138297303430.dylib'
17:55:52.693 [main] INFO  o.formats.importer.ImportCandidates - Depth: 4 Metadata Level: MINIMUM
17:55:52.696 [main] DEBUG o.formats.importer.cli.ErrorHandler - SCANNING: Depth:0 Num:    2 Tot:  n/a File: alczysko/testing
17:55:52.699 [main] DEBUG loci.common.Location - Location.mapFile: embedded-stream.raw -> null
17:55:52.700 [main] DEBUG loci.common.Location - Location.mapFile: embedded-stream.raw -> null
17:55:52.702 [main] DEBUG loci.formats.Memoizer - skipping memo: no directory given
17:55:52.717 [main] DEBUG loci.common.NIOByteBufferProvider - Using mapped byte buffer? false
17:55:52.720 [main] INFO  loci.formats.ImageReader - APNGReader initializing /Users/pwalczysko/testing/HDF_logo.svg.png
17:55:52.721 [main] DEBUG loci.formats.FormatHandler - APNGReader initializing /Users/pwalczysko/testing/HDF_logo.svg.png
17:55:52.721 [main] DEBUG loci.formats.FormatHandler - loci.formats.in.APNGReader.initFile(/Users/pwalczysko/testing/HDF_logo.svg.png)
17:55:52.731 [main] DEBUG loci.common.services.ServiceFactory - Loaded properties from: services.properties
17:55:52.732 [main] DEBUG loci.common.services.ServiceFactory - Added interface interface loci.formats.services.POIService and implementation class loci.formats.services.POIServiceImpl
17:55:52.733 [main] DEBUG loci.common.services.ServiceFactory - Added interface interface loci.formats.services.MDBService and implementation class loci.formats.services.MDBServiceImpl
17:55:52.733 [main] DEBUG loci.common.services.ServiceFactory - Added interface interface loci.formats.services.JPEGTurboService and implementation class loci.formats.services.JPEGTurboServiceImpl
17:55:52.734 [main] DEBUG loci.common.services.ServiceFactory - Added interface interface ome.codecs.services.LuraWaveService and implementation class ome.codecs.services.LuraWaveServiceImpl
17:55:52.734 [main] DEBUG loci.common.services.ServiceFactory - Added interface interface loci.common.services.S3ClientService and implementation class loci.common.services.S3ClientServiceImpl
17:55:52.735 [main] DEBUG loci.common.services.ServiceFactory - Added interface interface loci.formats.services.JAIIIOService and implementation class loci.formats.services.JAIIIOServiceImpl
17:55:52.736 [main] DEBUG loci.common.services.ServiceFactory - Added interface interface loci.formats.services.WlzService and implementation class loci.formats.services.WlzServiceImpl
17:55:52.736 [main] DEBUG loci.common.services.ServiceFactory - Added interface interface loci.formats.services.JHDFService and implementation class loci.formats.services.JHDFServiceImpl
17:55:52.737 [main] DEBUG loci.common.services.ServiceFactory - Added interface interface loci.formats.services.NetCDFService and implementation class loci.formats.services.NetCDFServiceImpl
17:55:52.738 [main] DEBUG loci.common.services.ServiceFactory - Added interface interface loci.formats.services.EXIFService and implementation class loci.formats.services.EXIFServiceImpl
17:55:52.738 [main] DEBUG loci.common.services.ServiceFactory - Added interface interface loci.formats.services.MetakitService and implementation class loci.formats.services.MetakitServiceImpl
17:55:52.739 [main] DEBUG loci.common.services.ServiceFactory - Added interface interface loci.formats.services.LuraWaveService and implementation class loci.formats.services.LuraWaveServiceImpl
17:55:52.740 [main] DEBUG loci.common.services.ServiceFactory - Added interface interface loci.formats.services.OMEXMLService and implementation class loci.formats.services.OMEXMLServiceImpl
17:55:52.741 [main] DEBUG loci.common.services.ServiceFactory - Added interface interface ome.codecs.services.JAIIIOService and implementation class ome.codecs.services.JAIIIOServiceImpl
17:55:52.741 [main] DEBUG loci.common.services.ServiceFactory - Added interface interface loci.formats.services.JPEGXRService and implementation class loci.formats.services.JPEGXRServiceImpl
17:55:52.778 [main] DEBUG loci.formats.Memoizer - start[1600966552702] time[75] tag[loci.formats.Memoizer.setId]
17:55:52.796 [main] DEBUG loci.common.Location - Location.mapFile: embedded-stream.raw -> null
17:55:52.796 [main] DEBUG loci.common.Location - Location.mapFile: embedded-stream.raw -> null
17:55:52.796 [main] DEBUG loci.common.Location - Location.mapFile: embedded-stream.raw -> null
17:55:52.796 [main] DEBUG loci.common.Location - Location.mapFile: embedded-stream.raw -> null
17:55:52.797 [main] DEBUG loci.formats.Memoizer - skipping memo: no directory given
17:55:52.797 [main] INFO  loci.formats.ImageReader - APNGReader initializing /Users/pwalczysko/testing/HDR5-2.png
17:55:52.797 [main] DEBUG loci.formats.FormatHandler - APNGReader initializing /Users/pwalczysko/testing/HDR5-2.png
17:55:52.797 [main] DEBUG loci.formats.FormatHandler - loci.formats.in.APNGReader.initFile(/Users/pwalczysko/testing/HDR5-2.png)
17:55:52.797 [main] DEBUG loci.formats.Memoizer - start[1600966552797] time[0] tag[loci.formats.Memoizer.setId]
17:55:52.797 [main] DEBUG loci.common.Location - Location.mapFile: embedded-stream.raw -> null
17:55:52.797 [main] DEBUG loci.common.Location - Location.mapFile: embedded-stream.raw -> null
17:55:52.798 [main] DEBUG o.formats.importer.cli.ErrorHandler - SCANNING: Depth:0 Num:    2 Tot:    2 File: alczysko/testing
17:55:52.799 [main] INFO  o.formats.importer.ImportCandidates - 2 file(s) parsed into 2 group(s) with 2 call(s) to setId in 101ms. (105ms total) [0 unknowns]
17:55:52.837 [main] INFO  ome.formats.OMEROMetadataStoreClient - Attempting initial SSL connection to outreach.openmicroscopy.org:4064
17:55:53.795 [main] INFO  ome.formats.OMEROMetadataStoreClient - Insecure connection requested, falling back
17:55:55.294 [main] INFO  ome.formats.OMEROMetadataStoreClient - Pinging session every 300s.
17:55:55.428 [main] INFO  ome.formats.OMEROMetadataStoreClient - Server: 5.6.2
17:55:55.428 [main] INFO  ome.formats.OMEROMetadataStoreClient - Client: 5.5.7
17:55:55.428 [main] INFO  ome.formats.OMEROMetadataStoreClient - Java Version: 1.8.0_222
17:55:55.428 [main] INFO  ome.formats.OMEROMetadataStoreClient - OS Name: Mac OS X
17:55:55.428 [main] INFO  ome.formats.OMEROMetadataStoreClient - OS Arch: x86_64
17:55:55.428 [main] INFO  ome.formats.OMEROMetadataStoreClient - OS Version: 10.13.6
17:55:56.015 [main] DEBUG ome.system.UpgradeCheck - Attempting to connect to http://upgrade.openmicroscopy.org.uk?version=5.5.7;os.name=Mac+OS+X;os.arch=x86_64;os.version=10.13.6;java.runtime.version=1.8.0_222-b10;java.vm.vendor=AdoptOpenJDK
17:55:56.329 [main] WARN  ome.system.UpgradeCheck - UPGRADE AVAILABLE:Please upgrade to 5.6.1. See http://downloads.openmicroscopy.org/latest/omero for the latest version.
17:55:56.331 [pool-2-thread-1] DEBUG ome.formats.importer.ImportLibrary - Main file: /Users/pwalczysko/testing/HDF_logo.svg.png
17:55:56.331 [pool-2-thread-1] DEBUG ome.formats.importer.ImportLibrary - Used files before:
17:55:56.331 [pool-2-thread-1] DEBUG ome.formats.importer.ImportLibrary - /Users/pwalczysko/testing/HDF_logo.svg.png
17:55:56.334 [pool-2-thread-1] INFO  o.f.i.cli.LoggingImportMonitor - FILESET_UPLOAD_PREPARATION
17:55:56.358 [pool-2-thread-1] DEBUG o.s.blitz.util.CurrentPlatform - recognized current operating system as being Apple Mac OS X
17:55:56.682 [pool-2-thread-1] INFO  o.f.i.cli.LoggingImportMonitor - FILESET_UPLOAD_START
17:55:56.780 [pool-3-thread-1] INFO  o.f.i.transfers.UploadFileTransfer - Transferring /Users/pwalczysko/testing/HDF_logo.svg.png...
17:55:56.838 [pool-3-thread-1] INFO  o.f.i.cli.LoggingImportMonitor - FILE_UPLOAD_STARTED: /Users/pwalczysko/testing/HDF_logo.svg.png
17:55:56.927 [pool-3-thread-1] DEBUG o.f.i.cli.LoggingImportMonitor - FILE_UPLOAD_BYTES uploaded: 0 of: 9881 bytes (ETA: 0' 0")
17:55:57.075 [pool-3-thread-1] DEBUG o.f.i.cli.LoggingImportMonitor - FILE_UPLOAD_BYTES uploaded: 9881 of: 9881 bytes (ETA: 0' 0")
17:55:57.129 [pool-3-thread-1] DEBUG o.f.importer.transfers.TransferState - trainer-1_2/Blitz-0-Ice.ThreadPool.Server-108/2020-09/24/17-55-56.508//HDF_logo.svg.png id=289337
17:55:57.129 [pool-3-thread-1] DEBUG o.f.importer.transfers.TransferState - checksums: client=6182bc8f3019df3b2a7f51ea0715933db59c27ff,server=6182bc8f3019df3b2a7f51ea0715933db59c27ff
17:55:57.129 [pool-3-thread-1] INFO  o.f.i.cli.LoggingImportMonitor - FILE_UPLOAD_COMPLETE: /Users/pwalczysko/testing/HDF_logo.svg.png
17:55:57.308 [pool-2-thread-1] INFO  o.f.i.cli.LoggingImportMonitor - FILESET_UPLOAD_END
17:55:57.410 [Ice.ThreadPool.Client-0] INFO  o.f.i.cli.LoggingImportMonitor - METADATA_IMPORTED Step: 1 of 5  Logfile: null
17:55:57.639 [Ice.ThreadPool.Client-0] INFO  o.f.i.cli.LoggingImportMonitor - PIXELDATA_PROCESSED Step: 2 of 5  Logfile: null
17:55:57.682 [pool-2-thread-1] INFO  o.f.i.cli.LoggingImportMonitor - IMPORT_STARTED Logfile: 289336
17:55:57.771 [Ice.ThreadPool.Client-1] INFO  o.f.i.cli.LoggingImportMonitor - THUMBNAILS_GENERATED Step: 3 of 5  Logfile: 289336
17:55:57.783 [Ice.ThreadPool.Client-0] INFO  o.f.i.cli.LoggingImportMonitor - METADATA_PROCESSED Step: 4 of 5  Logfile: 289336
17:55:57.795 [Ice.ThreadPool.Client-1] INFO  o.f.i.cli.LoggingImportMonitor - OBJECTS_RETURNED Step: 5 of 5  Logfile: 289336
17:55:57.935 [Ice.ThreadPool.Client-0] INFO  o.f.i.cli.LoggingImportMonitor - IMPORT_DONE Imported file: /Users/pwalczysko/testing/HDF_logo.svg.png
Image:58696
17:55:57.940 [Ice.ThreadPool.Client-0] DEBUG o.formats.importer.cli.ErrorHandler - Number of errors: 0
17:55:57.988 [pool-2-thread-1] DEBUG ome.formats.importer.ImportLibrary - Main file: /Users/pwalczysko/testing/HDR5-2.png
17:55:57.989 [pool-2-thread-1] DEBUG ome.formats.importer.ImportLibrary - Used files before:
17:55:57.989 [pool-2-thread-1] DEBUG ome.formats.importer.ImportLibrary - /Users/pwalczysko/testing/HDR5-2.png
17:55:57.989 [pool-2-thread-1] INFO  o.f.i.cli.LoggingImportMonitor - FILESET_UPLOAD_PREPARATION
17:55:58.245 [pool-2-thread-1] INFO  o.f.i.cli.LoggingImportMonitor - FILESET_UPLOAD_START
17:55:58.330 [pool-3-thread-1] INFO  o.f.i.transfers.UploadFileTransfer - Transferring /Users/pwalczysko/testing/HDR5-2.png...
17:55:58.389 [pool-3-thread-1] INFO  o.f.i.cli.LoggingImportMonitor - FILE_UPLOAD_STARTED: /Users/pwalczysko/testing/HDR5-2.png
17:55:58.477 [pool-3-thread-1] DEBUG o.f.i.cli.LoggingImportMonitor - FILE_UPLOAD_BYTES uploaded: 0 of: 9881 bytes (ETA: 0' 0")
17:55:58.612 [pool-3-thread-1] DEBUG o.f.i.cli.LoggingImportMonitor - FILE_UPLOAD_BYTES uploaded: 9881 of: 9881 bytes (ETA: 0' 0")
17:55:58.665 [pool-3-thread-1] DEBUG o.f.importer.transfers.TransferState - trainer-1_2/Blitz-0-Ice.ThreadPool.Server-108/2020-09/24/17-55-58.089//HDR5-2.png id=289340
17:55:58.666 [pool-3-thread-1] DEBUG o.f.importer.transfers.TransferState - checksums: client=6182bc8f3019df3b2a7f51ea0715933db59c27ff,server=6182bc8f3019df3b2a7f51ea0715933db59c27ff
17:55:58.666 [pool-3-thread-1] INFO  o.f.i.cli.LoggingImportMonitor - FILE_UPLOAD_COMPLETE: /Users/pwalczysko/testing/HDR5-2.png
17:55:58.837 [pool-2-thread-1] INFO  o.f.i.cli.LoggingImportMonitor - FILESET_UPLOAD_END
17:55:59.077 [pool-2-thread-1] INFO  o.f.i.cli.LoggingImportMonitor - IMPORT_STARTED Logfile: 289339
17:55:59.127 [Ice.ThreadPool.Client-1] INFO  o.f.i.cli.LoggingImportMonitor - PIXELDATA_PROCESSED Step: 2 of 5  Logfile: 289339
17:55:59.246 [Ice.ThreadPool.Client-0] INFO  o.f.i.cli.LoggingImportMonitor - THUMBNAILS_GENERATED Step: 3 of 5  Logfile: 289339
17:55:59.260 [Ice.ThreadPool.Client-1] INFO  o.f.i.cli.LoggingImportMonitor - METADATA_PROCESSED Step: 4 of 5  Logfile: 289339
17:55:59.272 [Ice.ThreadPool.Client-0] INFO  o.f.i.cli.LoggingImportMonitor - OBJECTS_RETURNED Step: 5 of 5  Logfile: 289339
17:55:59.374 [Ice.ThreadPool.Client-1] INFO  o.f.i.cli.LoggingImportMonitor - IMPORT_DONE Imported file: /Users/pwalczysko/testing/HDR5-2.png
Image:58697
17:55:59.375 [Ice.ThreadPool.Client-1] DEBUG o.formats.importer.cli.ErrorHandler - Number of errors: 0
17:55:59.426 [main] DEBUG ome.formats.OMEROMetadataStoreClient - Logout called, shutting keep alive down.
17:55:59.427 [main] DEBUG ome.formats.importer.ImportLibrary - Initializing log out
17:55:59.427 [main] DEBUG ome.formats.importer.ImportLibrary - Shutting down callback ome.formats.importer.ImportLibrary$ImportCallback@40cb698e
17:55:59.427 [main] DEBUG ome.formats.importer.ImportLibrary - Shutting down callback ome.formats.importer.ImportLibrary$ImportCallback@3382f8ae
17:55:59.427 [main] DEBUG ome.formats.importer.ImportLibrary - Shutting down fileset thread pool.
17:55:59.427 [main] DEBUG ome.formats.importer.ImportLibrary - Shutting down upload thread pool.
17:55:59.427 [main] DEBUG ome.formats.OMEROMetadataStoreClient - keepalive shut down.
17:55:59.742 [main] DEBUG ome.formats.OMEROMetadataStoreClient - closing client session.
17:55:59.865 [main] DEBUG ome.formats.OMEROMetadataStoreClient - client closed.
```

</details>

Reply:
Why don't you just do this, instead of the for-loop?
```
omero render set $DID renderingdef.yml
```

Reply:
Because it gives the way/idea of how to deal with the single images instead of the whole dataset if you wish so.
Maybe it needs more explanation and another example ?

Reply:
This feels like it should go in a more general "How to do useful things with the CLI" guide, since it is quite useful for a bunch of non-import scenarios. Certainly if you wanted to look-up how to do this, you wouldn't look on the import page.
I don't know if that place is in omero-guides or in OME docs - that's a wider discussion.

Reply:
@will-moore Cases where an output from one CLI command serves as an input for another is indeed a wider question. The problem there is that the code snippet would vary from case to case (I might be wrong there, but this is what I experienced).  The second problem is that it is at a whole not very easy, as the OMERO.cli is not always giving a streamlined output, and, the output is varying from one command to another. For example the ``omero hql...`` command output needed in cases which I worked with a conbindation of ``sed`` , ``grep`` , ``cut`` and similar commands to digest the output in such a way as to be consumed by, for example ``omero chown`` or ``omero chgrp``.
As you rightly remark, the documentation on this is completely absent, and the above might be the reason.  Changing and unifying the outputs and inputs of the commands will need a considerable effort in the OMERO.cli codebase.

@will-moore  I. will go for your suggestion with the $DID for this PR.

cc @sbesson

Reply:
fixed in https://github.com/ome/omero-guide-upload/pull/11/commits/e75d346846df75a975b3e41c0a975108c27fc3d0